### PR TITLE
feature/SPRIND-61_conformance-fixes

### PR DESCRIPTION
**Pull request by sanderPostma**

### CODE REVIEW

File: packages/key-utils/src/functions.ts [hunk 1]
```diff
@@ -28,7 +28,7 @@ export const logger = Loggers.DEFAULT.get('sphereon:key-utils')
  * @param kms. Optional KMS to use. If provided will be the returned name. Otherwise the default KMS will be returned
  */
 export const getKms = async (context: IAgentContext<any>, kms?: string): Promise<string> => {
-  if (kms) {
+  if (kms && kms !== 'default') {
```
Comment (nklomp) [inline packages/key-utils/src/functions.ts]: Huh. If I named a KMS default, then it should be returned. The below code already handles the case where we can determine a default KMS. So if we use a KMS (which we won't soon anymore) that does not have the option to determine a default we should throw an error. 

So what is this currently fixing? Because if we still use a Veramo Key management system, we should be using the Sphereon Key Management System

File: packages/jwt-service/src/functions/index.ts [hunk 2]
```diff
@@ -418,14 +420,18 @@ export const toJwsJsonGeneralWithIdentifiers = async (
...
         ? await resolveExternalJwkIdentifier({ identifier: args.jwk }, context)
         : await resolveExternalIdentifierFromJwsHeader(protectedHeader, context, args)
-      return { ...signature, identifier }
+      if (identifier !== undefined) {
+        return { ...signature, identifier }
+      }
+      return undefined
     })
-  )
+  )).filter(signature => signature !== undefined ) as JwsJsonSignatureWithIdentifier[]
```
Comment (nklomp) [inline packages/jwt-service/src/functions/index.ts:434]: This cannot be correct. We now all of a sudden return less objects than we should. Why is this change needed?

File: packages/jwt-service/src/types/IJwtService.ts [hunk 3]
```diff
@@ -164,4 +164,4 @@ export function isJwsJsonGeneral(jws: Jws): jws is JwsJsonGeneral {
   return typeof jws === 'object' && 'signatures' in jws
 }
 
-export const COMPACT_JWS_REGEX = /^([a-zA-Z0-9_=-]+)\.([a-zA-Z0-9_=-]+)?\.([a-zA-Z0-9_=-]+)$/
+export const COMPACT_JWS_REGEX = /^([a-zA-Z0-9_=-]+)\.([a-zA-Z0-9_=-]+)?(\.([a-zA-Z0-9_=-]+)?)?\.?$/
```
Comment (nklomp) [inline packages/jwt-service/src/types/IJwtService.ts:167]: Why not change the last + to * ?

**Review by nklomp — CHANGES_REQUESTED** [review]
Se eremarks. We cannot all of a sudden return less JWS recipients than inputs

File: packages/key-utils/src/functions.ts (hunk 1 above)
Comment (nklomp) [reply packages/key-utils/src/functions.ts]: Okay this has been reverted it seems

Comment (sanderPostma) [reply packages/key-utils/src/functions.ts]: yes reverted a while ago when I learned about keyManagerGetDefaultKeyManagementSystem()

File: packages/jwt-service/src/functions/index.ts (hunk 2 above)
Comment (sanderPostma) [reply packages/jwt-service/src/functions/index.ts:434]: I am possibly adding undefined elements on line 432. I need to return JwsJsonSignatureWithIdentifier[] , not (JwsJsonSignatureWithIdentifier | undefined)[]

File: packages/jwt-service/src/types/IJwtService.ts (hunk 3 above)
Comment (sanderPostma) [reply packages/jwt-service/src/types/IJwtService.ts:167]: I did not permit more than the extra dot.  Wouldn't changing the last + to * would capture the third part as empty string possibly increasing parts.length to 3?
I did not test that so I just added an optional dot.

Comment (nklomp) [reply packages/jwt-service/src/types/IJwtService.ts:167]: Even worse, I believe this regex playnly is wrong now as it will probably capture the wrong things in group 3 because you added two time a ? in there as well

Comment (nklomp) [reply packages/jwt-service/src/types/IJwtService.ts:167]: Just tested it quickly in an online regex tester, and indeed it is way of now

Comment (nklomp) [reply packages/jwt-service/src/types/IJwtService.ts]: Group 3 including a . all of a sudden. It starts accepting strings with 4 dots in there etc.

Comment (sanderPostma) [reply packages/jwt-service/src/types/IJwtService.ts]: I updated to be * and tested in regex101

Comment (nklomp) [reply packages/jwt-service/src/types/IJwtService.ts]: ```^([a-zA-Z0-9_=-]+)\.([a-zA-Z0-9_=-]+)?\.([a-zA-Z0-9_=-]+)?$```
If there is no payload so ey..ey then there is no group 2.  IF there is no signature so ey.ey. then there is no group 3. Seems to accomplish what you are after

Comment (nklomp) [reply packages/jwt-service/src/types/IJwtService.ts]: See my remark, because my initial suggestion returns an empty group 3, whilst the latest is in line with the initial expectation

Comment (sanderPostma) [reply packages/jwt-service/src/types/IJwtService.ts]: This does not allow extra groups like group 4
^([a-zA-Z0-9_=-]+)\.([a-zA-Z0-9_=-]+)(?:\.([a-zA-Z0-9_=-]+))?\.?$

However it does allow a 4th dot, but as soon as I type an extra char it stops matching
eyJ0eXAiOiJrYitqd3QiLCJhbGciOiJFUzI1NiJ9.eyJpYXQiOjE3MjcwODM2ODA4MzMsIm5vbmNlIjoiYjkzNzM2NzgtOTlhNi00NzYxLTk0Y2ItZWJmODMzZWQ0NTY4Iiwic2RfaGFzaCI6Il9icG5UMkdzNnZ4VTlxTWJhNkZDbnM5RGNVd2U5dXhldmlRUjNaY1d0a1EifQ.jlvRRDBXOmlVhDuhhJQ8GWF_LIAUtDEbUNWTNAtdpvBTNM2HsTT4QuM6zG7F2Jj1D0ri8qOzC1sGoslgNwcuYw.

Comment (nklomp) [reply packages/jwt-service/src/types/IJwtService.ts]: Yes which is invalid. Again see the regex in my last remark

Comment (sanderPostma) [reply packages/jwt-service/src/types/IJwtService.ts]: Your last remark says
^([a-zA-Z0-9_=-]+)\.([a-zA-Z0-9_=-]+)?\.([a-zA-Z0-9_=-]+)?$
That works, but for an unsigned one the second dot is now mandatory
eyJ0eXAiOiJrYitqd3QiLCJhbGciOiJFUzI1NiJ9.eyJpYXQiOjE3MjcwODM2ODA4MzMsIm5vbmNlIjoiYjkzNzM2NzgtOTlhNi00NzYxLTk0Y2ItZWJmODMzZWQ0NTY4Iiwic2RfaGFzaCI6Il9icG5UMkdzNnZ4VTlxTWJhNkZDbnM5RGNVd2U5dXhldmlRUjNaY1d0a1EifQ.
when I take the last dot it fails. Actually that's how it's in the spec, but we do loose some liency.

Comment (sanderPostma) [reply packages/jwt-service/src/types/IJwtService.ts]: tbc ^([a-zA-Z0-9_=-]+).([a-zA-Z0-9_=-]+)?.([a-zA-Z0-9_=-]+)?$ is now what's in the last commit

File: packages/jwt-service/src/functions/index.ts (hunk 2 above)
Comment (nklomp) [reply packages/jwt-service/src/functions/index.ts:434]: Yes, but the whole point is that you started introducing this. If for whatever reason the signature could be empty (which according to the spec it really can't for a jws json general AFAIK), then it still should return the identifiers, not all of a sudden filter everything out. 

But IMO this change reeks of something else going wrong when calling this function. So it should be solved outside of this function. If we really need this then we probably need something like an explicit param that allows the new behaviour (and only when that param is set)

Comment (sanderPostma) [reply packages/jwt-service/src/functions/index.ts]: resolved
